x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java

```diff
@@ -108,137 +108,185 @@ protected void masterOperation(
+                 * When creating the temporary rollup index, we copy the index.number_of_shards from source index,
+                 * and we set the index.number_of_replicas to 0, to avoid replicating the temp index.
+                 */
+                Settings.builder()
+                    .put(IndexMetadata.SETTING_INDEX_HIDDEN, true)
+                    .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, sourceIndexMetadata.getNumberOfShards())
+                    .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                    .build()
+            ).mappings(XContentHelper.convertToJson(BytesReference.bytes(mapping), false, XContentType.JSON));
+
+            // 2. Create hidden temporary rollup index
+            submitUnbatchedTask("create-rollup-index", new ClusterStateUpdateTask() {
                 @Override
                 public ClusterState execute(ClusterState currentState) throws Exception {
                     return metadataCreateIndexService.applyCreateIndexRequest(
                         currentState,
                         createIndexClusterStateUpdateRequest,
                         true,
-                        (builder, indexMetadata) -> builder.put(
-                            IndexMetadata.builder(indexMetadata)
-                                .settings(
-                                    Settings.builder()
-                                        .put(indexMetadata.getSettings())
-                                        .put(IndexMetadata.INDEX_ROLLUP_SOURCE_NAME.getKey(), sourceIndexName)
-                                        .put(IndexMetadata.INDEX_ROLLUP_SOURCE_UUID.getKey(), sourceIndexUuid)
-                                )
-                        )
+                        // Copy index metadata from source index to rollup index
+                        (builder, rollupIndexMetadata) -> builder.put(copyIndexMetadata(sourceIndexMetadata, rollupIndexMetadata))
                     );
                 }
 
                 public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
-                    // index created
-                    // 3.
+                    // 3. Temporary rollup index created. Run rollup indexer
+                    RollupIndexerAction.Request rollupIndexerRequest = new RollupIndexerAction.Request(
+                        tmpIndexName,
+                        request,
+                        dimensionFieldCaps.keySet().toArray(new String[0]),
+                        metricFieldCaps.keySet().toArray(new String[0])
+                    );
+
                     client.execute(RollupIndexerAction.INSTANCE, rollupIndexerRequest, ActionListener.wrap(indexerResp -> {
                         if (indexerResp.isCreated()) {
-                            // 4.
+                            // 4. Make temp index read-only
+                            UpdateSettingsRequest updateSettingsReq = new UpdateSettingsRequest(WRITE_BLOCKED_SETTINGS, tmpIndexName);
```

Comment:
I wonder if we should propagate parent task here as well.

Reply:
Fixed. Now it propagates parent task for all client calls